pkg/library/reconcilers/target_ref_reconciler.go

```diff
@@ -209,3 +213,144 @@ func (r *TargetRefReconciler) ReconcileGatewayPolicyReferences(ctx context.Conte
+		meta.RemoveStatusCondition(&gatewayDiffs[i].Status.Conditions, condition.Type)
+		if err := r.Client.Status().Update(ctx, gatewayDiffs[i].Gateway); err != nil {
+			return err
+		}
+		if err := r.updateInheritedGatewayCondition(ctx, condition, r.removeRouteCondition, gatewayDiffs[i].Gateway, policy, targetNetworkObject); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+type updateRouteConditionFunc func(ctx context.Context, condition metav1.Condition, route gatewayapiv1.HTTPRoute, gateway gatewayapiv1.Gateway, controllerName gatewayapiv1.GatewayController) error
+
+func (r *TargetRefReconciler) updateInheritedGatewayCondition(ctx context.Context, condition metav1.Condition, update updateRouteConditionFunc, gateway *gatewayapiv1.Gateway, policy kuadrantgatewayapi.Policy, targetNetworkObject client.Object) error {
+	logger, _ := logr.FromContext(ctx)
+	logger = logger.WithName("updateInheritedGatewayCondition").WithValues("condition", condition.Type, "gateway", client.ObjectKeyFromObject(gateway), "policy", policy)
+
+	if policy.PolicyClass() != kuadrantgatewayapi.InheritedPolicy {
+		return nil
+	}
+
+	var routes []gatewayapiv1.HTTPRoute
+
+	switch obj := targetNetworkObject.(type) {
+	case *gatewayapiv1.Gateway:
+		routes = r.FetchAcceptedGatewayHTTPRoutes(ctx, policy.TargetKey())
+	case *gatewayapiv1.HTTPRoute:
+		routes = append(routes, *obj)
+	}
+
+	var errs error
+
+	logger.V(1).Info("update inherited gateway condition", "#routes", len(routes))
+
+	for idx := range routes {
+		route := routes[idx]
+		if err := update(ctx, condition, route, *gateway, kuadrant.ControllerName); err != nil {
+			errs = errors.Join(errs, err)
+		}
+	}
+
+	return errs
+}
+
+func (r *TargetRefReconciler) addRouteCondition(ctx context.Context, condition metav1.Condition, route gatewayapiv1.HTTPRoute, gateway gatewayapiv1.Gateway, controllerName gatewayapiv1.GatewayController) error {
+	logger, _ := logr.FromContext(ctx)
+	logger = logger.WithName("addRouteCondition").WithValues("condition", condition.Type, "route", client.ObjectKeyFromObject(&route), "gateway", client.ObjectKeyFromObject(&gateway))
+	i := utils.Index(route.Status.RouteStatus.Parents, func(p gatewayapiv1.RouteParentStatus) bool {
+		return *p.ParentRef.Kind == ("Gateway") &&
+			p.ControllerName == controllerName &&
+			((p.ParentRef.Namespace == nil && route.GetNamespace() == gateway.GetNamespace()) || string(*p.ParentRef.Namespace) == gateway.GetNamespace()) &&
+			string(p.ParentRef.Name) == gateway.GetName()
+	})
+	if i < 0 {
+		logger.V(1).Info("cannot find parent status, creating new one")
+		route.Status.RouteStatus.Parents = append(route.Status.RouteStatus.Parents, gatewayapiv1.RouteParentStatus{
+			ControllerName: controllerName,
+			ParentRef: gatewayapiv1.ParentReference{
+				Kind:      ptr.To(gatewayapiv1.Kind(gateway.Kind)),
+				Name:      gatewayapiv1.ObjectName(gateway.GetName()),
+				Namespace: ptr.To(gatewayapiv1.Namespace(gateway.GetNamespace())),
+			},
```

Comment:
This will create in the route a parent status block with `controllerName: kuadrant.io/policy-controller` and `targetRef` pointing to the affected gateway.

One can only assume what the actual `controllerName` of the gateway is, but this information is omitted here.

Should we instead reuse the same parent status block of the gateway controller and not have one specifically for Kuadrant?